File: services/games/internal/mathobjects/rect.go
package mathobjects

import "math"

type Rect struct {
	Position Vector2[float64] `json:"position"`
	Size     Vector2[float64] `json:"size"`
}

func CreateRect(x, y, w, h float64) Rect {
	return Rect{
		Position: Vector2[float64]{x, y},
		Size:     Vector2[float64]{w, h},
	}
}

func (rect *Rect) Intersects(rect2 Rect) bool {
	if rect.Position.X+rect.Size.X <= rect2.Position.X ||
		rect.Position.X >= rect2.Position.X+rect2.Size.X ||
		rect.Position.Y+rect.Size.Y <= rect2.Position.Y ||
		rect.Position.Y >= rect2.Position.Y+rect2.Size.Y {
		return false
	}

	return true
}

func (rect *Rect) SetPosition(x, y float64) {
	rect.Position.X = x
	rect.Position.Y = y
}

func (rect *Rect) SetSize(x, y float64) {
	rect.Size.X = x
	rect.Size.Y = y
}

func (rect *Rect) Extend(extX, extY float64) {
	rect.Size.X += math.Abs(extX)
	rect.Size.Y += math.Abs(extY)

	if extX < 0 {
		rect.Position.X -= math.Abs(extX)
	}
	if extY < 0 {
		rect.Position.Y -= math.Abs(extY)
	}
}

func (rect *Rect) GetPosition() Vector2[float64] {
	return rect.Position
}

func (rect *Rect) GetSize() Vector2[float64] {
	return rect.Size
}

File: storageFiles/typescript/gameassets/room/rect.ts
function lerp(a: number, b: number, alpha: number): number{
    return a + alpha * (b - a);
}

interface AbstractRect {
    position: {
        x: number,
        y: number,
    }
    size: {
        x: number,
        y: number,
    }
}

function isAbstractRect(obj: any): obj is AbstractRect {
    return "position" in obj && "size" in obj
}

class Rect {
    protected position: Vector2;
    protected size: Vector2;

    constructor(abstractRect?: AbstractRect){
        this.position = new Vector2(0, 0);
        this.size = new Vector2(0, 0);

        if(abstractRect){
            this.setPosition(abstractRect.position.x, abstractRect.position.y);
            this.setSize(abstractRect.size.x, abstractRect.size.y);
        }
    }

    setPosition(x: number, y: number){
        this.position.setPosition(x, y);
    }

    setSize(x: number, y: number){
        this.size.setPosition(x, y);
    }

    extend(extX: number, extY: number){
        this.size.x += Math.abs(extX)
        this.size.y += Math.abs(extY)
        if (extX < 0){
            this.position.x -= Math.abs(extX)
        }
        if (extY < 0){
            this.position.y -= Math.abs(extY)
        }
    }

    containsPoint(x: number, y: number){
        let pos = this.position;
        let size = this.size;

        if(
        x >= pos.x && x <= pos.x + size.x &&
        y >= pos.y && y <= pos.y + size.y
        ){
            return true;
        }
        return false;
    }

    intersects(rect: Rect){
        if (this.position.x+this.size.x <= rect.position.x ||
        this.position.x >= rect.position.x+rect.size.x ||
        this.position.y+this.size.y <= rect.position.y ||
        this.position.y >= rect.position.y+rect.size.y){
            return false
        }

        return true
    }

    getPosition(){
        return new Vector2(this.position.x, this.position.y);
    }

    getSize(){
        return new Vector2(this.size.x, this.size.y);
    }
}